### PR TITLE
BUG: Set check_exact to true if dtype is int

### LINKED ISSUE #55882: BUG: assert_series_equal not raising on unequal series?
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
left = pd.Series([1577840521123000])
right = pd.Series([1577840521123543])
pd.testing.assert_series_equal(left, right)
```

### Issue Description

The above doesn't raise

### Expected Behavior

It should raise

### Installed Versions

<details>

In [4]: pd.show_versions()
/home/marcogorelli/tmp/.venv/lib/python3.10/site-packages/_distutils_hack/__init__.py:33: UserWarning: Setuptools is replacing distutils.
  warnings.warn("Setuptools is replacing distutils.")

INSTALLED VERSIONS
------------------
commit              : 51f3d03087414bdff763d3f45b11e37b2a28ea84
python              : 3.10.12.final.0
python-bits         : 64
OS                  : Linux
OS-release          : 5.10.102.1-microsoft-standard-WSL2
Version             : #1 SMP Wed Mar 2 00:30:59 UTC 2022
machine             : x86_64
processor           : x86_64
byteorder           : little
LC_ALL              : None
LANG                : en_GB.UTF-8
LOCALE              : en_GB.UTF-8

pandas              : 2.2.0.dev0+541.g51f3d03087
numpy               : 1.25.1
pytz                : 2023.3
dateutil            : 2.8.2
setuptools          : 67.6.1
pip                 : 23.1.2
Cython              : None
pytest              : 7.3.1
hypothesis          : 6.82.4
sphinx              : None
blosc               : None
feather             : None
xlsxwriter          : None
lxml.etree          : None
html5lib            : None
pymysql             : None
psycopg2            : None
jinja2              : 3.1.2
IPython             : 8.16.1
pandas_datareader   : None
bs4                 : 4.12.2
bottleneck          : None
dataframe-api-compat: None
fastparquet         : 2023.7.0
fsspec              : 2023.6.0
gcsfs               : None
matplotlib          : 3.7.1
numba               : 0.58.1
numexpr             : None
odfpy               : None
openpyxl            : None
pandas_gbq          : None
pyarrow             : 12.0.1
pyreadstat          : None
python-calamine     : None
pyxlsb              : None
s3fs                : None
scipy               : 1.11.0
sqlalchemy          : 2.0.20
tables              : None
tabulate            : None
xarray              : None
xlrd                : None
zstandard           : None
tzdata              : 2023.3
qtpy                : None
pyqt5               : None

</details>

### PR DESCRIPTION
**Pull request by parthi-siva**
- [x] closes #55882 
- [x] [Tests added and passed]
- [x] All [code checks passed]
- [ ] Added [type annotations]
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Comment by parthi-siva** [comment]
@MarcoGorelli not sure where to add test cases for the change.

File: pandas/_testing/asserters.py [hunk 1]
```diff
@@ -921,6 +921,14 @@ def assert_series_equal(
         else:
             assert_attr_equal("dtype", left, right, obj=f"Attributes of {obj}")
 
+    if (
+        is_integer_dtype(left.dtype)
+        and is_integer_dtype(right.dtype)
+        and isinstance(left._values, type(right._values))
+        and isinstance(right._values, type(left._values))
```
Comment (MarcoGorelli) [inline pandas/_testing/asserters.py:928]: just for my understanding, why do we need these two lines?

Comment (parthi-siva) [reply pandas/_testing/asserters.py:928]: Sure.
 In this test case pandas/tests/series/test_constructors.py::TestSeriesConstructors::test_constructor_maskedarray:572
one assertion is failing.

```
    def test_constructor_maskedarray(self):
        data = ma.masked_all((3,), dtype=float)
        result = Series(data)
        expected = Series([np.nan, np.nan, np.nan])
        tm.assert_series_equal(result, expected)
    
        data[0] = 0.0
        data[2] = 2.0
        index = ["a", "b", "c"]
        result = Series(data, index=index)
        expected = Series([0.0, np.nan, 2.0], index=index)
        tm.assert_series_equal(result, expected)
    
        data[1] = 1.0
        result = Series(data, index=index)
        expected = Series([0.0, 1.0, 2.0], index=index)
        tm.assert_series_equal(result, expected)
    
        data = ma.masked_all((3,), dtype=int)
        result = Series(data)
        expected = Series([np.nan, np.nan, np.nan], dtype=float)
        tm.assert_series_equal(result, expected)
    
        data[0] = 0
        data[2] = 2
        index = ["a", "b", "c"]
        result = Series(data, index=index)
        expected = Series([0, np.nan, 2], index=index, dtype=float)
        tm.assert_series_equal(result, expected)
    
        data[1] = 1
        result = Series(data, index=index)
        expected = Series([0, 1, 2], index=index, dtype=int)
>       tm.assert_series_equal(result, expected)
E       AssertionError: Series are different
E       
E       Series classes are different
E       [left]:  MaskedArray
E       [right]: ndarray
```

Even though `left.dtype `and` right.dtype `are same their type are different. So here are setting check_exact to true only when the types of left and right are same.

**Comment by parthi-siva** [comment]
> Well I'm relieved that CI is still green 😄 Can you update the docs for this method to note that `rtol`, `atol` and `check_exact` don't take effect for int dtype?
> 
> Regarding the check - not totally sure about the check, will take a closer look next week

Sure. Updated the documentation. Added note under `check_exact`. Since we are setting it to True in case of int dtype. Also, documentation for `rtol` and `atol` states it is applicable only when check_exact is False I assumed that it is sufficient to add not under `check_exact`

**Comment by MarcoGorelli** [comment]
I think it should always be exact, except for floats?

You should also check for extensions arrays, and include a whatsnew

I've added a commit anyway, hope it's ok - do the changes look alright to you?

**Comment by parthi-siva** [comment]
Yeah, make sense.

Some testcases are failing in some builds. Is that alright?

**Comment by MarcoGorelli** [comment]
Looks like another bug was hiding because of the check_exact default 😄  https://github.com/pandas-dev/pandas/issues/56136

For now let's xfail the relevant test

**Review by MarcoGorelli — APPROVED** [review]
Looks good to me - although I did add a commit, so leaving open a bit in case others have objections

**Comment by parthi-siva** [comment]
Sure Marco

File: pandas/tests/extension/base/methods.py [hunk 2]
```diff
@@ -331,7 +332,8 @@ def test_fillna_length_mismatch(self, data_missing):
             data_missing.fillna(data_missing.take([1]))
 
     # Subclasses can override if we expect e.g Sparse[bool], boolean, pyarrow[bool]
-    _combine_le_expected_dtype: Dtype = np.dtype(bool)
+    # _combine_le_expected_dtype: Dtype = np.dtype(bool)
```
Comment (mroeschke) [inline pandas/tests/extension/base/methods.py:335]:
```suggestion
```

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @parthi-siva and @MarcoGorelli

File: pandas/tests/series/test_constructors.py [hunk 3]
```diff
@@ -572,7 +572,10 @@ def test_constructor_maskedarray(self):
         data[1] = 1
         result = Series(data, index=index)
         expected = Series([0, 1, 2], index=index, dtype=int)
-        tm.assert_series_equal(result, expected)
+        with pytest.raises(AssertionError, match="Series classes are different"):
+            # TODO should this be raising at all?
+            # https://github.com/pandas-dev/pandas/issues/56131
+            tm.assert_series_equal(result, expected)
```
Comment (jorisvandenbossche) [inline pandas/tests/series/test_constructors.py:578]: This is a different issue than reported in https://github.com/pandas-dev/pandas/issues/56131 (there is also no `check_dtype=False` here)

This doesn't involve different data types (it's all numpy dtypes), and the two objects created here seemingly are the same:

```python
data = np.ma.masked_all((3,), dtype=int)
data[0] = 0
data[1] = 1
data[2] = 2
index = ["a", "b", "c"]
result = Series(data, index=index)
expected = Series([0, 1, 2], index=index, dtype=int)
```
```
In [29]: result
Out[29]: 
a    0
b    1
c    2
dtype: int64

In [30]: expected
Out[30]: 
a    0
b    1
c    2
dtype: int64

In [31]: result.dtype == expected.dtype
Out[31]: True
```

But apparently we have a bug in the Series constructor that preserves the masked array as underlying value if it has no masked elements:

```
In [32]: result.values
Out[32]: 
masked_array(data=[0, 1, 2],
             mask=[False, False, False],
       fill_value=999999)

In [33]: expected.values
Out[33]: array([0, 1, 2])
```

That seems like a separate bug we should solve.

**Comment by parthi-siva** [comment]
> Looks like another bug was hiding because of the check_exact default 😄 #56136
> 
> For now let's xfail the relevant test

@jorisvandenbossche yeah.. there was another issues because of this

**Pull request merged**